File: examples/python/tutorials/03_first_algorithm.py
from pathlib import Path

import lynxes as lx

ROOT = Path(__file__).resolve().parents[3]
GRAPH_PATH = ROOT / "examples" / "data" / "example_simple.gf"


def main() -> None:
    # This toy repository assumes a local checkout. Full Lynxes docs live in the
    # upstream Lynxes repository.
    graph = lx.read_gf(GRAPH_PATH)

    # Eager algorithms run immediately on a materialized GraphFrame. The output
    # is still frame-shaped, so it can be inspected without leaving the Lynxes
    # surface or inventing a separate result wrapper.
    path = graph.shortest_path("alice", "charlie")
    print(f"graph: {GRAPH_PATH.name}")
    print(f"shortest path: {path}")

    ranks = graph.pagerank()
    print("pagerank preview:")
    print(ranks.head(5, sort_by="pagerank", descending=True))


if __name__ == "__main__":
    main()

File: examples/rust/tutorials/03_first_algorithm.rs
use std::fs;

use lynxes::{parse_gf, PageRankConfig, ShortestPathConfig};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // These examples assume you are running from a repository checkout and can
    // read the shared example graphs under examples/data.
    let simple = fs::read_to_string("examples/data/example_simple.gf")?;
    let simple_graph = parse_gf(&simple)?.to_graph_frame()?;

    // Eager algorithms execute immediately and return engine-native outputs that
    // can be inspected without switching to a different abstraction first.
    let path = simple_graph.shortest_path("alice", "charlie", &ShortestPathConfig::default())?;
    println!("shortest path: {:?}", path);

    let ranks = simple_graph.pagerank(&PageRankConfig::default())?;
    println!("pagerank rows: {}", ranks.len());
    Ok(())
}
